### ISSUE TITLE
Supervisor crashes due to OOM when livepushing

### ISSUE THREAD

**Issue by xginn8** [balena-io/balena-supervisor#1088]
Seen on a device running supervisor 10.3.2:

```
Sep 10 17:54:54 e655ba7 resin-supervisor[4726]: [error]   Error streaming local log output: Error: connect EAGAIN /var/run/balena-engine.sock
Sep 10 17:54:54 e655ba7 resin-supervisor[4726]: [error]         at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1097:14)
Sep 10 17:54:54 e655ba7 resin-supervisor[4726]: [error]   Error streaming local log output: Error: connect EAGAIN /var/run/balena-engine.sock
Sep 10 17:54:54 e655ba7 resin-supervisor[4726]: [error]         at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1097:14)
Sep 10 17:54:54 e655ba7 resin-supervisor[4726]: [error]   Error streaming local log output: Error: connect EAGAIN /var/run/balena-engine.sock
Sep 10 17:54:54 e655ba7 resin-supervisor[4726]: [error]         at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1097:14)
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: <--- Last few GCs --->
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: [1:0x1627f80]  1289124 ms: Mark-sweep 1350.5 (1454.9) -> 1348.3 (1454.9) MB, 433.9 / 0.0 ms  (average mu = 0.116, current mu = 0.034) allocation failure scavenge might not succeed
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: [1:0x1627f80]  1289644 ms: Mark-sweep 1351.1 (1454.9) -> 1349.6 (1455.4) MB, 507.0 / 0.0 ms  (average mu = 0.070, current mu = 0.025) allocation failure scavenge might not succeed
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: <--- JS stacktrace --->
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]:  1: 0x7373f0 node::Abort() [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]:  2: 0x73743c  [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]:  3: 0x8c182e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]:  4: 0x8c1a62 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]:  5: 0xc156f2  [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]:  6: 0xc15832  [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]:  7: 0xc24d72 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]:  8: 0xc256eb v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]:  9: 0xc27dfd v8::internal::Heap::AllocateRawWithLigthRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 10: 0xc27e52 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 11: 0xbf7476 v8::internal::Factory::AllocateRawWithImmortalMap(int, v8::internal::PretenureFlag, v8::internal::Map*, v8::internal::AllocationAlignment) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 12: 0xbfe5c0 v8::internal::Factory::NewRawOneByteString(int, v8::internal::PretenureFlag) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 13: 0xbfe861 v8::internal::Factory::NewStringFromOneByte(v8::internal::Vector<unsigned char const>, v8::internal::PretenureFlag) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 14: 0xbfeb50 v8::internal::Factory::NewStringFromUtf8(v8::internal::Vector<char const>, v8::internal::PretenureFlag) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 15: 0x8cd7b6 v8::String::NewFromUtf8(v8::Isolate*, char const*, v8::NewStringType, int) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 16: 0x7fd5ec80dcaf node_sqlite3::Statement::RowToJS(std::vector<node_sqlite3::Values::Field*, std::allocator<node_sqlite3::Values::Field*> >*) [/usr/src/app/node_modules/sqlite3/lib/binding/node-v64-linux-x64/node_sqlite3.node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 17: 0x7fd5ec80e658 node_sqlite3::Statement::Work_AfterAll(uv_work_s*) [/usr/src/app/node_modules/sqlite3/lib/binding/node-v64-linux-x64/node_sqlite3.node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 18: 0x83ed1a  [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 19: 0x842db4  [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 20: 0x851848  [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 21: 0x843650 uv_run [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 22: 0x741855  [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 23: 0x740433 node::Start(int, char**) [node]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 24: 0x7fd5ef3edb6b __libc_start_main [/lib/libc.so.6]
Sep 10 17:55:06 e655ba7 resin-supervisor[4726]: 25: 0x6fc19a _start [node]
Sep 10 17:55:08 e655ba7 systemd[1]: resin-supervisor.service: Main process exited, code=exited, status=139/n/a
Sep 10 17:55:08 e655ba7 systemd[1]: resin-supervisor.service: Failed with result 'exit-code'.
Sep 10 17:55:19 e655ba7 systemd[1]: resin-supervisor.service: Service RestartSec=10s expired, scheduling restart.
Sep 10 17:55:19 e655ba7 systemd[1]: resin-supervisor.service: Scheduled restart job, restart counter is at 1.
```

Specifically, I was running the following livepush command on BoB (intel-nuc device type):

```
DEBUG=1 balena push --registry-secrets registry.json e655ba7.local --service devices
```

These OOMs are reproducible when livepushing to BoB, and this may be a red herring, but always coincide with the following messages flooding the log:

```
Error streaming local log output: Error: connect EAGAIN /var/run/balena-engine.sock
```

cc @hedss since it's BoB

**Comment by hedss** [comment]
I hope the balenaEngine side is fixed as @robertgzr pushed a change for EAGAIN, but it sounds like something's not getting cleaned up in the Supervisor...

**Comment by xginn8** [comment]
I've reproduced this as far back as 10.2.2, and @CameronDiver has reproduced on balenaOS 2.41 + supervisor 10.2.2

**Comment by robertgzr** [comment]
> I hope the balenaEngine side is fixed as @robertgzr pushed a change for EAGAIN
>

The engine fix for the EAGAIN issue was part of [18.9.8](https://github.com/balena-os/balena-engine/blob/master/CHANGELOG.md#v1898) which came with [2.41.0](https://github.com/balena-os/meta-balena/blob/master/CHANGELOG.md#v2410)

**Comment by CameronDiver** [comment]
Closed by https://github.com/balena-io/balena-supervisor/pull/1099
